### ISSUE TITLE
Need help parsing an array of objects

### ISSUE THREAD

**Issue by reFX-Mike** [simdjson/simdjson#2063]
I have this valid ```layout_CRT.json``` file and can't seem to get the objects and arrays within:

```
[
	{ "monitor": [
		{ "id": "monitor",		"type": "toggle",		"label": "monitor"			},
		{ "id": "profile",		"type": "selector",		"label": "collection"		},
		{ "id": "overlay",		"type": "selector",		"label": "overlay"			},
		{ "id": "zoom",			"type": "toggleSlider",	"label": "zoom"				}
	] },

	{ "crt": [
		{ "id": "system",		"type": "multi",		"label": "system",		"choices": "PAL, NTSC"	},
		{ "type": "spacer" },
		{ "id": "brightness",	"type": "slider",		"icon": "brightness"		},
		{ "id": "contrast",		"type": "slider",		"icon": "contrast"			},
		{ "id": "saturation",	"type": "slider",		"icon": "saturation"		},
		{ "type": "spacer" },
		{ "id": "overscan",		"type": "toggleSlider",	"label": "overscan"			},
		{ "type": "spacer" },
		{ "id": "emulation",	"type": "toggle",		"label": "CRT emulation"	},
		{ "type": "spacer" },
		{ "id": "curve",		"type": "toggleSlider",	"label": "curve"			},
		{ "id": "bleed",		"type": "toggleSlider",	"label": "bleed"			},
		{ "id": "vignette",		"type": "toggleSlider",	"label": "vignette"			},
		{ "id": "scanlines",	"type": "toggleSlider",	"label": "scanlines"		},
		{ "id": "gridlines",	"type": "toggleSlider",	"label": "gridlines"		},
		{ "id": "glow",			"type": "toggleSlider",	"label": "glow"				},
		{ "id": "flicker",		"type": "toggleSlider",	"label": "flicker"			},
		{ "id": "noise",		"type": "toggleSlider",	"label": "noise"			}
	] }
]
```

I want to loop over the outermost array, get the name of the array-objects (first "monitor", then "crt") and also loop through their children:

```
    auto	json = simdjson::padded_string::load ( std::string ( path ) );
    auto	parser = simdjson::ondemand::parser {};
    auto	doc = parser.iterate ( json );
    auto	rootArray = doc.get_array ();

    for ( auto node : rootArray )
    {
        // Get name of object here
       // and second for-range loop to get child-elements here
    }
```

The reason I used arrays is that I need the order of the elements preserved. This is for a very basic layout system, but I'm stumped (again) by simdjson.

Also is it possible to have empty ```{}``` child-objects (I've used ```"type": "spacer"``` as a work-around for now)?

**Comment by lemire** [comment]
> The reason I used arrays is that I need the order of the elements preserved

At least in simdjson, the order of the keys is preserved. I think that's generally true.


You could just do...

```json
{
	"monitor": [{
			"id": "monitor",
			"type": "toggle",
			"label": "monitor"
		},
		{
			"id": "profile",
			"type": "selector",
			"label": "collection"
		},
		{
			"id": "zoom",
			"type": "toggleSlider",
			"label": "zoom"
		}
	],
	"crt": [{
			"id": "system",
			"type": "multi",
			"label": "system",
			"choices": "PAL, NTSC"
		},
		{
			"type": "spacer"
		},
		{
			"id": "noise",
			"type": "toggleSlider",
			"label": "noise"
		}
	]
}
```

Nevertheless, with the schema you have now, you can do something like this...
```C++
#include "simdjson.h"
#include <iostream>
using namespace simdjson;

int main() {
  padded_string padded_input_json = R"([
	{ "monitor": [
		{ "id": "monitor",		"type": "toggle",		"label": "monitor"			},
		{ "id": "profile",		"type": "selector",		"label": "collection"		},
		{ "id": "overlay",		"type": "selector",		"label": "overlay"			},
		{ "id": "zoom",			"type": "toggleSlider",	"label": "zoom"				}
	] },

	{ "crt": [
		{ "id": "system",		"type": "multi",		"label": "system",		"choices": "PAL, NTSC"	},
		{ "type": "spacer" },
		{ "id": "brightness",	"type": "slider",		"icon": "brightness"		},
		{ "id": "contrast",		"type": "slider",		"icon": "contrast"			},
		{ "id": "saturation",	"type": "slider",		"icon": "saturation"		},
		{ "type": "spacer" },
		{ "id": "overscan",		"type": "toggleSlider",	"label": "overscan"			},
		{ "type": "spacer" },
		{ "id": "emulation",	"type": "toggle",		"label": "CRT emulation"	},
		{ "type": "spacer" },
		{ "id": "curve",		"type": "toggleSlider",	"label": "curve"			},
		{ "id": "bleed",		"type": "toggleSlider",	"label": "bleed"			},
		{ "id": "vignette",		"type": "toggleSlider",	"label": "vignette"			},
		{ "id": "scanlines",	"type": "toggleSlider",	"label": "scanlines"		},
		{ "id": "gridlines",	"type": "toggleSlider",	"label": "gridlines"		},
		{ "id": "glow",			"type": "toggleSlider",	"label": "glow"				},
		{ "id": "flicker",		"type": "toggleSlider",	"label": "flicker"			},
		{ "id": "noise",		"type": "toggleSlider",	"label": "noise"			}
	] }
])"_padded;
  auto parser = ondemand::parser{};
  auto doc = parser.iterate(padded_input_json);
  auto root_array = doc.get_array();
  // the root should be an object, not an array, but that's the JSON we are
  // given.
  for (ondemand::object node : root_array) {
    // We know that we are going to have just one element in the object.
    for (auto field : node) {
      std::cout << "\n\ntop level:" << field.key() << std::endl;
      // You can get a proper std::string_view for the key with:
      // std::string_view key = field.unescaped_key();
      // and second for-range loop to get child-elements here
      for (auto inner_object : field.value().get_array()) {
        for (auto inner_field : inner_object.get_object()) {
          std::cout << '"' << inner_field.key() << "\" : " << inner_field.value()
                    << ", ";
          // You can get proper std::string_view for the key and value with:
          // std::string_view inner_key = field.unescaped_key();
          // std::string_view value_str = field.value();
        }
        std::cout << std::endl;
      }
      // You can break here if you only want just the first element.
      // break;
    }
  }
}
```

If you simplify your scheme as I describe above, then you will peel off one layer in the processing code.

My code sample assumes that you are using exceptions, if not, you need to do a bit more work for the error handling.

**Comment by lemire** [comment]
> Also is it possible to have empty {} child-objects (I've used "type": "spacer" as a work-around for now)?

Yes. The simdjson library supports the full JSON specification, including empty objects and empty arrays.

**Comment by reFX-Mike** [comment]
> > Also is it possible to have empty {} child-objects (I've used "type": "spacer" as a work-around for now)?
> 
> Yes. The simdjson library supports the full JSON specification, including empty objects and empty arrays.

And how would I check for that in the inner-object loop? Can you provide a brief example for that as well, please?

**Comment by reFX-Mike** [comment]
> I think that's generally true.

Only in simdjson. The JSON format explicitly mentions that there are no guarantees and that any parsing code can reorder the objects in literally any order it likes. e.g. if you use a std::map or even an unordered_map, the order in which you then read it back from is totally random. The only workaround is to use arrays or numbered keys.

**Comment by lemire** [comment]
> Only in simdjson

Many JSON APIs guarantee the order, not just simdjson. But you are correct that the spec itself makes no guarantee. 

Another approach is to use an array, with interleaved values (string, object, string, object...).

> And how would I check for that in the inner-object loop?

It is difficult for me to know what you are doing with the code (i.e., I expect you don't print it out) but you mean that you want to check whether you have iterated at least once through the loop? You can do it with a flag such as...

```C++
 size_t is_empty = true;
 for (auto inner_field : inner_object.get_object()) {
   is_empty = false;
   std::cout << '"' << inner_field.key() << "\" : " << inner_field.value()
                    << ", ";
 }
 if(is_empty) { std::cout << "it is empty" << std::endl; }
````

**Comment by reFX-Mike** [comment]
Imagine inside the “CRT” object, that one, or multiple, of the children are simply empty objects. I want to detect that and then skip that object. Only if the object contains any fields, then I want to process it.

Basically the objects with the type “spacer” should really just be empty objects, without the type field.

**Comment by lemire** [comment]
@reFX-Mike You could use iterators, like so (look for `if (i == inner_object.end()) {`) 

```C++
#include "simdjson.h"
#include <iostream>
using namespace simdjson;

int main() {
  padded_string padded_input_json = R"([
	{ "monitor": [
		{ "id": "monitor",		"type": "toggle",		"label": "monitor"			},
		{ "id": "profile",		"type": "selector",		"label": "collection"		},
		{ "id": "overlay",		"type": "selector",		"label": "overlay"			},
		{ "id": "zoom",			"type": "toggleSlider",	"label": "zoom"				}
	] },

	{ "crt": [
		{ "id": "system",		"type": "multi",		"label": "system",		"choices": "PAL, NTSC"	},
		{ "type": "spacer" },
		{ "id": "brightness",	"type": "slider",		"icon": "brightness"		},
		{ "id": "contrast",		"type": "slider",		"icon": "contrast"			},
		{ "id": "saturation",	"type": "slider",		"icon": "saturation"		},
		{ "type": "spacer" },
		{ "id": "overscan",		"type": "toggleSlider",	"label": "overscan"			},
		{ "type": "spacer" },
		{ "id": "emulation",	"type": "toggle",		"label": "CRT emulation"	},
		{ "type": "spacer" },
		{ "id": "curve",		"type": "toggleSlider",	"label": "curve"			},
		{ "id": "bleed",		"type": "toggleSlider",	"label": "bleed"			},
		{ "id": "vignette",		"type": "toggleSlider",	"label": "vignette"			},
		{ "id": "scanlines",	"type": "toggleSlider",	"label": "scanlines"		},
		{ "id": "gridlines",	"type": "toggleSlider",	"label": "gridlines"		},
		{ "id": "glow",			"type": "toggleSlider",	"label": "glow"				},
		{ "id": "flicker",		"type": "toggleSlider",	"label": "flicker"			},
		{ "id": "noise",		"type": "toggleSlider",	"label": "noise"			},
    {}
	] }
])"_padded;
  auto parser = ondemand::parser{};
  auto doc = parser.iterate(padded_input_json);
  auto root_array = doc.get_array();
  // the root should be an object, not an array, but that's the JSON we are
  // given.
  for (ondemand::object node : root_array) {
    // We know that we are going to have just one element in the object.
    for (auto field : node) {
      std::cout << "\n\ntop level:" << field.key() << std::endl;
      // You can get a proper std::string_view for the key with:
      // std::string_view key = field.unescaped_key();
      // and second for-range loop to get child-elements here
      for (ondemand::object inner_object : field.value()) {
        auto i = inner_object.begin();
        if (i == inner_object.end()) {
          std::cout << "empty object" << std::endl;
          continue;
        } else {
          for (; i != inner_object.end(); ++i) {
            auto inner_field = *i;
            std::cout << '"' << inner_field.key()
                      << "\" : " << inner_field.value() << ", ";
            // You can get proper std::string_view for the key and value with:
            // std::string_view inner_key = field.unescaped_key();
            // std::string_view value_str = field.value();
          }
        }
        std::cout << std::endl;
      }
      // You can break here if you only want just the first element.
      // break;
    }
  }
}
```

Does that help?

**Comment by reFX-Mike** [comment]
Unfortunately, is does not work. If I check do this:

```
auto	y = 0;
for ( simdjson::ondemand::object node : rootArray )
{
	for ( auto field : node )
	{
		auto	section = std::string_view ( field.unescaped_key () );
		DBG ( std::string ( section ) );

		for ( auto inner_object : field.value ().get_array () )
		{
			if ( inner_object.begin () != inner_object.end () )
			{
				auto	id = std::string_view ( inner_object[ "id" ] );

				DBG ( "   " << std::string ( id ) );
			}
			y += 24;
		}
	}
}
```

The loop terminates early as soon as the first empty object is hit, it throws an exception of ```NO_SUCH_FIELD: The JSON field referenced does not exist in this object.````

Here is the update json-file:

```
[
	{ "monitor": [
		{ "id": "monitor",		"type": "toggle",		"label": "monitor"			},
		{ "id": "profile",		"type": "selector",		"label": "collection"		},
		{ "id": "overlay",		"type": "selector",		"label": "overlay"			},
		{ "id": "zoom",			"type": "toggleSlider",	"label": "zoom"				}
	] },

	{ "crt": [
		{ "id": "system",		"type": "multi",		"label": "system",		"choices": "PAL, NTSC"	},
		{},
		{ "id": "brightness",	"type": "slider",		"icon": "brightness"		},
		{ "id": "contrast",		"type": "slider",		"icon": "contrast"			},
		{ "id": "saturation",	"type": "slider",		"icon": "saturation"		},
		{},
		{ "id": "overscan",		"type": "toggleSlider",	"label": "overscan"			},
		{},
		{ "id": "emulation",	"type": "toggle",		"label": "CRT emulation"	},
		{},
		{ "id": "curve",		"type": "toggleSlider",	"label": "curve"			},
		{ "id": "bleed",		"type": "toggleSlider",	"label": "bleed"			},
		{ "id": "vignette",		"type": "toggleSlider",	"label": "vignette"			},
		{ "id": "scanlines",	"type": "toggleSlider",	"label": "scanlines"		},
		{ "id": "gridlines",	"type": "toggleSlider",	"label": "gridlines"		},
		{ "id": "glow",			"type": "toggleSlider",	"label": "glow"				},
		{ "id": "flicker",		"type": "toggleSlider",	"label": "flicker"			},
		{ "id": "noise",		"type": "toggleSlider",	"label": "noise"			}
	] }
]
```

**Comment by reFX-Mike** [comment]
OK. If I change this one line

```for ( auto inner_object : field.value ().get_array () )```
to
```for ( simdjson::ondemand::object inner_object : field.value ().get_array () )```

then, it parses all the way through, skipping the empty objects. But at the very end, when the parsing is successfully finished, it emmits an illegal instruction. I found that simdjson.cpp does this if something went wrong, but I've printed out all outer keys and inner IDs I'm interested in (into the debug output) and it all seems to work correctly. How do I find out what went wrong?

Or do I have to consume every single field? Or do something else?
